### PR TITLE
Refactor and improve model estimator tool

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR, while on the surface changes no values, will work towards improving the estimator tool by allowing us to explicitly state the various types of memory one could expect. 

Specifically we can now differentiate on the four memory workloads during training:
1. Model + Optimizer
2. Gradients
3. Backward pass
4. Optimizer step

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@SunMarc

### CODE REVIEW

File: src/accelerate/commands/estimate.py [hunk 1]
```diff
@@ -212,6 +212,40 @@ def estimate_command_parser(subparsers=None):
...
+    fp16_size = bytes // 2
+
+    if mixed_precision == "float32":
+        memory_sizes["model"] = fp32_size
+        memory_sizes["gradients"] = fp32_size
+        memory_sizes["optimizer"] = fp32_size * 2
+        memory_sizes["step"] = fp32_size * 4
+    elif mixed_precision in ("float16", "bfloat16") or mixed_precision == "fp8" and msamp_config is None:
+        # With mixed precision training, the model has weights stored
+        # in FP16 and FP32
```
Comment (SunMarc) [inline src/accelerate/commands/estimate.py:239]: I think that the FP8 case is not handled properly. WE should introduce `fp8_size` maybe . Also I think it is missing parentheses around `mixed_precision == "fp8" and msamp_config is None `.

**Review by SunMarc — COMMENTED** [review]
Thanks for adding this. I left a comment !

File: src/accelerate/commands/estimate.py (hunk 1 above)
Comment (muellerzr) [reply src/accelerate/commands/estimate.py:239]: There is no real memory savings with FP8 rn, it is the exact same as with float16 :)

**Review by SunMarc — APPROVED** [review]
LGTM !

**Pull request merged**